File: dev/Neuron.js
"use strict"

class Neuron {
    
    constructor (importedData) {
        if (importedData) {
            this.imported = true
            this.weights = importedData.weights || []
            this.bias = importedData.bias
        }
    }

    init (size, {adaptiveLR, activationConfig, eluAlpha}={}) {

        this.deltaWeights = this.weights.map(v => 0)

        switch (adaptiveLR) {
            
            case "gain":
                this.biasGain = 1
                this.weightGains = [...new Array(size)].map(v => 1)
                this.getWeightGain = i => this.weightGains[i]
                this.setWeightGain = (i,v) => this.weightGains[i] = v
                break

            case "adagrad":
            case "rmsprop":
            case "adadelta":
                this.biasCache = 0
                this.weightsCache = [...new Array(size)].map(v => 0)
                this.getWeightsCache = i => this.weightsCache[i]
                this.setWeightsCache = (i,v) => this.weightsCache[i] = v

                if (adaptiveLR=="adadelta") {
                    this.adadeltaBiasCache = 0
                    this.adadeltaCache = [...new Array(size)].map(v => 0)
                    this.getAdadeltaCache = i => this.adadeltaCache[i]
                    this.setAdadeltaCache = (i,v) => this.adadeltaCache[i] = v
                }
                break

            case "adam":
                this.m = 0
                this.v = 0
                break
        }

        if (activationConfig=="rrelu") {
            this.rreluSlope = Math.random() * 0.001
            
        } else if (activationConfig=="elu") {
            this.eluAlpha = eluAlpha
        }
    }

    getWeight (i) {
        return this.weights[i]
    }

    setWeight (i,v) {
        this.weights[i] = v
    }

    getDeltaWeight (i) {
        return this.deltaWeights[i]
    }

    setDeltaWeight (i,v) {
        this.deltaWeights[i] = v
    }
}

typeof window=="undefined" && (exports.Neuron = Neuron)

File: dev/cpp/Neuron.cpp

void Neuron::init (int netInstance, int weightsCount) {

    Network* net = Network::getInstance(netInstance);

    dropped = false;

    switch (net->updateFnIndex) {
        case 1: // gain
            biasGain = 1;
            for (int i=0; i<weightsCount; i++) {
                weightGain.push_back(1);
            }
            break;
        case 2: // adagrad
        case 3: // rmsprop
        case 5: // adadelta
        case 6: // momentum
            biasCache = 0;
            for (int i=0; i<weightsCount; i++) {
                weightsCache.push_back(0);
            }

            if (net->updateFnIndex == 5) {
                adadeltaBiasCache = 0;
                for (int i=0; i<weightsCount; i++) {
                    adadeltaCache.push_back(0);
                }
            }
            break;

        case 4: // adam
            m = 0;
            v = 0;
            break;
    }

    if (net->activation == &NetMath::lrelu<Neuron>) {
        lreluSlope = net->lreluSlope;
    } else if (net->activation == &NetMath::rrelu<Neuron>) {
        rreluSlope = ((double) rand() / (RAND_MAX)) * 0.001;
    } else if (net->activation == &NetMath::elu<Neuron>) {
        eluAlpha = net->eluAlpha;
    }
}